### ISSUE TITLE
[REQUEST]: 

### ISSUE THREAD

**Issue by Artem-B** [compiler-explorer/compiler-explorer#4328]
### Is your feature request related to a problem? Please describe

It appears that by default for CUDA compilation w/ NVCC in CE now shows *host* code and that the GPU PTX/SASS is shown in the new "device" pane.

I think it's a regression in usability.

- examination of the host code is rarely useful for CUDA compilation.
- In order to examine PTX and SASS simultaneously, we now have to open another pair of compiler + device viewer windows, so we end up with four windows in total, two of them with redundant and useless host code. Previously we only needed the two windows with SASS (via output->compile-to-binary) and PTX output which was the primary compiler window output.

### Describe the solution you'd like

I would propose to keep PTX as the default compiler output and push the host code output under "device" viewer.

### Describe alternatives you've considered

- ignore host code altogether. It worked fine so far. C++ compilation does that job just fine and it's almost never needs to be correlated w/ the GPU code.
- Allow output variant switching from within the main compiler output window (I.e. no separate device window). 
  - We sort of do it already in form of "output->compile to binary"
  - This has the benefit that user does not need to be stuck with an extra window they don't need.
  - The downside is that it's a binary switch, while we may potentially have multiple output kinds (host + {PTX, SASS} * N GPUs)
- Allow multiple device windows. This has the downside of having the ever-present compiler output windows with the host code around. It's better than what we have now, but that useless host window should not be there, unless user needs it.

### Additional context

Not applicable

**Comment by Artem-B** [comment]
@trxcllnt WDYT?

**Comment by partouf** [comment]
Thanks for the feedback.

The device window has really been an effort to simplify and streamline the code and user-experience, not just for CUDA, but any compiler that produces code for multiple targets. It has also opened doors for easier execution, as we don't need to differentiate between binary and executable anymore.

As to your points:
* Keep PTX as default - this will not solve anything, as the code for other devices will still have to go somewhere.
* Switching from the main compiler - This will clutter up the UI with things it will not need for most other compilers. And it's a hard thing to manage in terms of both developing that because it will inevitably result in bugs.
* Revert to using binary mode - That's definitely a no-go, it will lead to complicated code and offers no option for people that want to see something other than PTX.

* Allow multiple device windows - This is probably something that we can do, and probably already should have done.

**Comment by Artem-B** [comment]
I agree that we have multiple dimensions we can slice source-to-binary pipeline(s) across, so devices, as a way to add one such dimension is a good thing. 

> Allow multiple device windows - This is probably something that we can do, and probably already should have done.

That may be work. I guess I could downsize the host output window and tuck it under the source code where it would be out of the way. If we could have some sort of other way to quickly get that windows out of the way (and restore back as it carries compiler options I'll need to tweak) that would be even better.

Speaking of multiple device windows -- would IR/AST output would also be eventually delegated to 'Device' view?
E.g. let's suppose that we compile a CUDA source for sm50 and sm60, under the hood the compilation will look roughly like this (* caveat, the pipeline may be somewhat different, depending on options):
```
a.cu +-> AST[sm_50] -> IR[sm_50] -> PTX[sm_50] -> SASS[sm_50] -\
     \-> AST[sm_60] -> IR[sm_60] -> PTX[sm_60] -> SASS[sm_60] --- [fatbin] -\
      \-> AST[host] -----------------------------------------------------------> IR[host] -> ASM[host]
```
Currently PTX/SASS are selectable from the device view, but conceptually we should be able to have similar functionality for other stages of compilation pipelines.

Which brings us to the fact that different compilers have different intermediate stages. Dumping them all under 'device' may be a bit of a mess, too. Perhaps instead of calling it a 'device' view, we should allow each compiler to declare its internal 'pipeline geometry' or 'intermediate results' with relevant names and parsers/highlighters, and let user select the ones they want or need?

In a way, in case of CUDA, the host output is not in any way special, compared to GPU-side assembly. What makes it special in CE is that it's tied to compilation options. IMO, compiler options and selection of the intermediate results the user may want to see should be independent. Device output makes the step in the right direction, but I think it does not go far enough.

As a strawman, here's roughly what I would want to see for CUDA.
- source windows -- as usual.
- "compilation" - specifies compiler and its command line.
- "result selection". It would generally depend on the options above. I.e. for CUDA compilation targeting multiple GPUs there will be more things to select. Still the number of entries will likely be limited and we could get by with a multiple choice list, with commonly used entries chosen by default. It could actually be part of the compiler options window with the list hidden by default and expanded if/when user needs to. E.g. once options are set, CE could run clang with `-###` and populate the list of all intermediate outputs it would produce during the real compilation. NVCC would do the same and produce its own set of outputs. Bonus points that it would allow exposing other compiler-specific results in a more uniform way. E.g. AST/IR I've mentioned above. If we want to get fancy, we could even show user a graph of how those inputs are connected. Or, possibly allow them to parametrize particular intermediate result (e.g. give me IR before and after IR pass X, which is something that's very common for those who work on LLVM. Other compilers will have fixed pipelines that will not depend on compiler options and could skip such pre-run, or, even default to providing a single always-on choice -- ASM.
- intermediate result window(s), based on the results selected above.

**Comment by pauleonix** [comment]
While I like to have the option to see the host code, I didn't directly catch that, especially since there is host assembly being identified with regions in device only functions and kernels.

Another sharp edge is that one cannot use the device pane in a "Diff View". So when one tries to compare the PTX or SASS as someone who used Compiler-Explorer for that before, one might only realize too late that one is actually comparing host assembly, especially if the Diff doesn't show any difference and one doesn't take a closer look at the instructions.

**Comment by jrhemstad** [comment]
I think the new changes are a net positive, but, as always, there is room for improvement. 

> As a strawman, here's roughly what I would want to see for CUDA.
>     * source windows -- as usual.
>     * "compilation" - specifies compiler and its command line.
>     * "result selection". 
>     * intermediate result window(s), based on the results selected above.

I'd be supportive of something like this.

I think our current problem stems from the fact that specifying compiler options/libraries/etc are intrinsically tied to the host code assembly view. Ideally these would be able to move independently such that I could configure compilation/libraries/etc and this would be independent of the various assembly/IR/AST output. 

As to how exactly this is accomplished from a UI perspective, I will defer to my betters :slightly_smiling_face:. 

> Which brings us to the fact that different compilers have different intermediate stages. Dumping them all under 'device' may be a bit of a mess, too. Perhaps instead of calling it a 'device' view, we should allow each compiler to declare its internal 'pipeline geometry' or 'intermediate results' with relevant names and parsers/highlighters, and let user select the ones they want or need?

With gcc/clang in the C++ Language mode, there are unique options under the "Add New" for things like the AST or gcc graph output. Presumably we could do a similar thing and not everything would need to go under the device-view? 

![latest-screenshot](https://user-images.githubusercontent.com/15221289/204398773-a0a8105a-362a-46a9-af7a-c595482c0e54.png)

**Comment by partouf** [comment]
Optimization, Preprocessor, AST, LLVM IR, LLVM Opt - all of these really do "extra things", extra compilations to get the results for that specific function. The device view does not need anything, it's automatically supplied with the compilation or extra compiler arguments.

So they're really not related to a compilation pipeline in that sense at the moment.

We could of course change all that and make everything optional or whatever, but I'm not sure that is worth the effort or that it will solve anything.

**Comment by jrhemstad** [comment]
> all of these really do "extra things", extra compilations to get the results for that specific function. The device view does not need anything, it's automatically supplied with the compilation or extra compiler arguments.

Ah, okay. That makes sense as a split between what goes under "Add new" vs the device view.

**Comment by Artem-B** [comment]
> Optimization, Preprocessor, AST, LLVM IR, LLVM Opt - all of these really do "extra things", extra compilations to get the results for that specific function. The device view does not need anything, it's automatically supplied with the compilation or extra compiler arguments.

Yes, no,  "it depends", and not necessarily in that order. :-)

E.g. preprocessor and optimized IR will be dumped by clang if compiled w/ `-save-temps`, so getting them is (nearly) as free as getting assembly and object code  (at least if what NVCC compilation does is representative of the intent of the design). The fact that we already do the extra steps necessary for AST and other outputs that need that, means that it's already doable, so rearranging how we get to choose them is probably not going to change things drastically. You can think of a it as a more flexible way to populate "Add New" menu.

In any case, that would be answering the "how" part of the design, but we haven't figured out "what" we want/should do or, maybe, even "whether" we should change anything at all.

> I think our current problem stems from the fact that specifying compiler options/libraries/etc are intrinsically tied to the host code assembly view. 

I'd say that it's a common (and often implicit) assumption, but it's not universally true. CUDA/HIP would be one example where the host code is way less interesting than the GPU code for the vast majority of use cases. 

> As to how exactly this is accomplished from a UI perspective, I will defer to my betters 🙂.

That, and all the machinery under the hood.
It's likely that making live easier for a niche set of users like CUD/HIP is not going to be worth it, but I'd like to figure out a way to minimize the impact of the "new way" on the practical usability for CUDA users like myself. CE is an invaluable tool for me, too, and I want to keep it that way.

> With gcc/clang in the C++ Language mode, there are unique options under the "Add New" for things like the AST or gcc graph output. Presumably we could do a similar thing and not everything would need to go under the device-view?

That would not help with the host code window being ever stuck on the screen for no good reason.
Nor would it change the fact that one would need to go and manually open the window they actually do need, though device view auto-open does help with that to some degree, 

I guess, I'd be OK with something as simple as being able to choose what goes into the compiler result window by default. I.e. a checkbox "make device output the default compiler result" somewhere easily accessible would do. Basically I want the most common use case (for me) to be as easy to use as possible, without unnecessary stuff cluttering the display.
